specs/capri/Monitoring Design.md

```diff
@@ -0,0 +1,742 @@
+
+>   *annotations:*
+
+>   *summary: "Instance {{ \$labels.instance }} is low on disk space"*
+
+>   *description: "{{ \$labels.instance }} has only {{ \$value }}% free."*
+
+1.  The Anomaly detector detects an anomaly, raises a custom alert to
+    AlertManager using the REST API interface of AlertManager
+
+2.  Example script to send Alert using AlertManager API
+
+>   *alertsender.sh*
+```sh
+alerts1='[
+        {
+                "labels": {
+                        "alertname": "DiskRunningFull",
+                        "dev": "sda1",
+                        "instance": "example1"
+                },      
+                "annotations": {
+                        "info": "The disk sda1 is running full",
+                        "summary": "please check the instance example1"
+                }
+        }
+]'
+curl -XPOST -d"\$alerts1" http://\<alertmanagerIP\>:9093/api/v1/alerts
+```
+
+### Metric Format and Data Collection
+------------
+
+Metric notation in Prometheus is a metric name and a set of labels.
+
+\<metric name\>{\<label name\>=\<label value\>, ...}
+
+Prometheus Naming standards:- <https://prometheus.io/docs/practices/naming/>
+
+Prometheus Data model:- <https://prometheus.io/docs/concepts/data_model/>
+
+Metric collector in OpenSDS collects from different targets. Data model
+represents the structure of a metric in hotpot project. Data exporter
+(Prometheus push/ Kafka consumer) and metric Controller can use this generic
+type to derive function logic.
+
+metricName can be formed by concatenating some values of struct field.
+
+## Example: 
+
+*node_disk_read_bytes_total* , where
+
+node-\> Source of collection
+
+disk-\> component which the metric is collected from
+
+read-\> name of the KPI
+
+bytes-\> unit
+
+total-\> if it is a aggregated metric
+
+### Here is the proposed metric data model and mapping to Prometheus data model
+```go
+type Metric struct {
+
+        // Following are the labels associated with Metric, same as Prometheus labels
+
+        // Example: {device="dm-0",instance="121.244.95.60:12419",job="prometheus"}
+
+        // Instance ID -\> volumeID/NodeID  
+        instanceID string
+
+        // instance name -\> volume name / node name etc.  
+        instanceName string
+
+        // job -\> Prometheus/openSDS  
+        job string
+
+        /*associator - Some metric would need specific fields to relate components.
+        Use case could be to query volumes of a particular pool. Attaching the related
+        components as labels would help us to form promQl query efficiently.
+
+        Example: node_disk_read_bytes_total{instance="121.244.95.60"}
+
+        Above query will respond with all disks associated with node 121.244.95.60
+
+        Since associated components vary, we will keep a map in metric struct to denote
+        the associated component type as key and component name as value
+
+        Example: associator[pool]=pool1 */
+
+        associator map[string]string
```

Comment:
Is that ok if all fields in MetricSpec use lower case?

Reply:
should be okay, any specifc concern please ?